software/CONFIGURATION.md

```diff
@@ -53,3 +53,53 @@ shows the default configuration:
 ```
 
 - `volts_per_octave` must be one of `1.0` (Eurorack standard) or `1.2` (Buchla standard)
+
+
+# Accessing config members in Python code
+
+The firmware converts the JSON file into a `ConfigSettings` object, where the JSON keys are converted
+to Python attributes.  The attribute names are the same as their JSON strings, but converted to upper case (so as to
+appear as constants in the code) and with any alphanumeric characters replaced with `_` characters.  If the key starts
+with a number, a `K_` prefix is added. e.g.:
+
+- `language` -> `.LANGUAGE`
+- `display_channel` -> `.DISPLAY_CHANNEL`
+- `2pi` -> `K_2PI`
```

Comment:
`K_` doesn't stand for anything; it's just a common naming convention for constants in a lot of programming languages. Because JSON is more permissive with its keys (any string is a valid key) this extra bit of processing was added instead of raising an exception or ignoring the key.

Reply:
What do you think of us having a rule that config point names can't start with a number or symbol? it seems like a reasonable restriction and it will avoid the confusion that the weird `K_` prefix might cause.

Reply:
I'd prefer not limiting valid JSON syntax like that. We're already mangling the name by replacing non-alphanumeric characters with underscores & case-changes, so further mangling with a prefix if needed doesn't feel like it's crossing a line we haven't already crossed. I honestly doubt many people will have keys that start with numbers anyway.

Constants & macros starting with `k` is a pretty common thing across programming languages (referred to as "Hungarian notation), so I can't see this adding much confusion if it does ever come up.

Reply:
> I'd prefer not limiting valid JSON syntax like that. 

I'm not sure that I understand this reasoning. There are plenty of syntactically valid JSON files that are not valid europi config files. We are using a subset of JSON to define our config file format, and we can constrain subset as we see fit.

>We're already mangling the name by replacing non-alphanumeric characters with underscores & case-changes, 

I'd be quite happy to restrict the config keys to follow the python 'constants' variable name convention, and remove all of the mangling entirely.

> Constants & macros starting with `k` is a pretty common thing across programming languages (referred to as "Hungarian notation"), so I can't see this adding much confusion if it does ever come up.

In my experience Hungarian notation is not common in python.

@roryjamesallen @awonak do either of you have strong feelings here?

Reply:
I'm inclined to agree with @mjaskula and enforce that JSON keys should adhere to [Python Pep8](https://peps.python.org/pep-0008/#naming-conventions) variable naming conventions (ideally restricted to [constant format](https://peps.python.org/pep-0008/#constants)) because those keys will be converted into Python variable names. It's better to keep them as consistent as possible and not add any additional convention.

Reply:
Fair enough. I've made the requested changes and updated the documentation accordingly. The `ConfigSettings` will now raise a `ValueError` if an invalid key is provided.

Reply:
I'm more hesitant about case preservation simply because we've already got config JSON files in the wild. If we change the keys we invalidate existing files, which could cause issues for some users.

Reply:
Yeah, that part does suck. Right now, early in the feature's life (while the number of config files is very low), is probably the best time to make that change. So, personally I still think it's a good idea, but if everyone else is opposed to that kind of change for the users then I won't push it anymore.

Reply:
I've rarely seen JSON files using all-caps keys. That's not to say it's invalid, but the general consensus seems to be that JSON should either be in `snake_case` or `camelCase`, not `ALL_CAPS`. No idea why, but in my experience that's the consensus across most of the professional world.

On the other hand, the trend for read-only constants being in all caps is also pretty well established. So I don't really see a simple case-change as a deal-breaker.

Given the fact that we've already got config files in the wild using lower-case, as was originally established for EuroPi's configuration files, and that the industry generally seems to consider lower-case JSON to be "proper" I disagree that we should force the config files to use upper-case keys.